File: split/endtokenlist.pas
procedure endtokenlist;
begin
  if curinput.indexfield >= 3 then
  begin
    if curinput.indexfield <= 4 then
      flushlist(curinput.startfield)
    else
    begin
      deletetokenref(curinput.startfield);
      if curinput.indexfield = 5 then
        while paramptr > curinput.limitfield do
        begin
          paramptr := paramptr - 1;
          flushlist(paramstack[paramptr]);
        end;
    end;
  end
  else if curinput.indexfield = 1 then
    if alignstate > 500000 then
      alignstate := 0
    else
      fatalerror(595);
  begin
    inputptr := inputptr - 1;
    curinput := inputstack[inputptr];
  end;
  begin
    if interrupt <> 0 then
      pauseforinstructions;
  end;
end;

File: cpp/src/endtokenlist.cpp
#include "endtokenlist.h"
#include "flushlist.h"
#include "deletetokenref.h"
#include "erreur.h"
#include "pauseforinstructions.h"
#include "popinput.h"

void endtokenlist(void)
{
	if (token_type >= backed_up)
	{
		if (token_type <= inserted)
			flushlist(start);
		else
		{
			deletetokenref(start);
			if (token_type == macro)
				for (;paramptr > param_start; paramptr--)
					flushlist(paramstack[paramptr]);
		}
	}
	else 
		if (token_type == u_template)
			if (alignstate > 500000)
				alignstate = 0;
			else
				fatalerror("(interwoven alignment preambles are not allowed)");
	pop_input();
	check_interrupt();
}
